operatorcourier/push.py

```diff
@@ -42,15 +47,33 @@ def _push_to_registry(self, namespace, repository, release, bundle, auth_token):
         headers = {'Content-Type': 'application/json', 'Authorization': auth_token}
         json = {'blob': bundle, 'release': release, "media_type": "helm"}
         r = requests.post(push_uri, json=json, headers=headers)
-        if r.status_code != 200:
-            logger.error(r.text)
-
-            msg = 'Failed to get error details'
-            try:
-                error = r.json()['error']
-                msg = "Quay.io answered: '{}' ({})".format(
-                    error['code'], error['message'])
-            except Exception:
-                pass
-
-            raise ValueError("Failed to push to app registry: {}".format(msg))
+
+        if r.status_code == 200:
+            return
+
+        logger.error(r.text)
+        msg = "Failed to push to app registry: {}"
+
+        try:
```

Comment:
You may not even get here, IIRC Connection errors are raise directly by `requests.post()`, only HTTP status from reponse is raised by `raise_for_status`

Reply:
Would it then be safe to assume that if `requests.post()` doesn't fail, the response will always contain valid JSON content from  Quay? Or should I keep using `OpCourierQuayCommunicationError` for that and use `OpCourierQuayError` for exceptions raised by `requests.post()`?

Reply:
The API unfortunately does not have a public spec and the source code is not yet open. However, I believe it is a fair assumption that, if requests.post() on that endpoint returns some kind of success that there will always be a json response, even if that response is empty. I haven't seen any behavior otherwise.

Reply:
It needs something like this:
```python
try:
    r = request.post(.....)
except RequestsException as e:  # or Exception
   raise OpCourierQuayCommunicationError(str(e))
else:
   if r.status_code != 200:
     try:
        r_json = r.json()
     except Exception:
        r_json = {}
     raise OpCourierQuayErrorResponse('something', code, r_json)
```

Reply:
That looks reasonable. The only potential problem could that raising `OpCourierQuayCommunicationError` instead of `RequestException` would break backward compatibility. Of course, `OpCourierQuayError` could inherit from `RequestException` as well, though that would still break code like this: 
```python
try:
    api.build_verify_and_push()
except ValueError:
    # OpCourierQuayError gets caught here because it inherits from both
except RequestException:
    # and this doesn't happen
```
I'm probably making things unnecessarily complicated, but I'd still like to get your thoughts on it @MartinBasti  @kevinrizza

Reply:
@chmeliik @MartinBasti Personally I think anything where `r.status_code != 200` should except one error type that inherits from `ValueError` and anything where `requests.post()` throws an exception itself should except another error type that inherits `RequestException`. To me I thought that was the whole point of separating these error types.

Reply:
Ok, it could be done in this way